### PR TITLE
DRPC tests improvements

### PR DESCRIPTION
**Pull request by ELENAGER**
Several improvements in DRPC tests:
1. Moving common helpers to separate file
2. Use retryOnConflict function during update k8s resources
3. Make code more clean and clear

Fixes: https://github.com/RamenDR/ramen/issues/906

### CODE REVIEW

File: controllers/drplacementcontrol_controller_test.go [hunk 1]
```diff
@@ -1658,101 +1645,38 @@ func deleteDRPolicySync() {
...
-			continue
-		}
-
-		Expect(err).NotTo(HaveOccurred())
-
-		break
-	}
+	latestDRCluster := getLatestDRCluster(cluster)
+	latestDRCluster.Spec.ClusterFence = ""
+	updateDRClusterParameters(latestDRCluster)
```
Comment (nirs) [inline controllers/drplacementcontrol_controller_test.go:1697]: Looks like good change for its own commit - other changes in this file are not related to this change, right?

File: controllers/drplacementcontrol_controller_test.go [hunk 2]
```diff
@@ -1658,101 +1645,38 @@ func deleteDRPolicySync() {
...
-		err := k8sClient.Update(context.TODO(), latestDRCluster)
-		if errors.IsConflict(err) {
-			localRetries++
-
-			continue
-		}
-
-		Expect(err).NotTo(HaveOccurred())
-
-		break
```
Comment (nirs) [inline controllers/drplacementcontrol_controller_test.go:1727]: Replacing the retry loop with our "standard" retry loop looks good...

File: controllers/drplacementcontrol_controller_test.go [hunk 3]
```diff
@@ -1658,101 +1645,38 @@ func deleteDRPolicySync() {
...
-		if drClusterFencedCondition == nil {
-			return false
-		}
-
-		if drClusterFencedCondition.ObservedGeneration != latestDRCluster.Generation {
-			return false
-		}
-
-		return drClusterFencedCondition.Status == metav1.ConditionFalse
-	}, timeout, interval).Should(BeTrue(), "failed to update DRCluster on time")
```
Comment (nirs) [inline controllers/drplacementcontrol_controller_test.go:1744]: ...but I'm not sure about this change - replacing the `Eventually()` block with a call to `drclusterConditionExpectEventually()`.

Maybe it is correct, but this is exactly what we should not do - mix unrelated changes in the same commit. This makes it much harder to understand the change both during review of when trying to fix an issue later.

**Review by nirs — COMMENTED** [review]
The last commit seems to have 3 unrelated changes that are better done in a separate commit.

Also the commit messages are not helpful, they do not describe the change and why it is needed.

File: controllers/drplacementcontrol_controller_test.go (hunk 1 above)
Comment (ELENAGER) [reply controllers/drplacementcontrol_controller_test.go:1697]: This commit description is: "Use updateDRClusterParameters in functions fenceCluster, unfenceCluster and resetdrCluster". I supposed, that if the action is the same - replace code with using helper - I must not create the separate commit for every function. If you want to separate - we need 3 different commits here

File: controllers/drplacementcontrol_controller_test.go (hunk 3 above)
Comment (ELENAGER) [reply controllers/drplacementcontrol_controller_test.go:1744]: The change itself is correct, we are doing the same in other places in code (for example, in drclustercontroller_test.go), but it can be separated into another commit

Comment (ELENAGER) [reply controllers/drplacementcontrol_controller_test.go:1744]: Move to a separate commit

File: controllers/utils_test.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
+// SPDX-FileCopyrightText: The RamenDR authors
+// SPDX-License-Identifier: Apache-2.0
```
[the remark this answers is not in the document]
Comment (nirs) [reply controllers/utils_test.go:2]: Seems that there no good way to add utils module for tests which is not part of the application and the least worst way is to have a xxx_test.go module for the utils. But utils_test.go is confusing with the controllers/utils package. How about naming this `controllers_test.go` - using same name of the package doe snot clash with anything and can be a convention for other packages.

File: controllers/drplacementcontrol_controller_test.go [hunk 5]
```diff
@@ -691,32 +690,28 @@ func deleteNamespaceMWsFromAllClusters(namespace string) {
 }
 
 func setDRPCSpecExpectationTo(action rmn.DRAction, preferredCluster, failoverCluster string) {
-	localRetries := 0
-	for localRetries < updateRetries {
-		latestDRPC := getLatestDRPC()
+	drpcLookupKey := types.NamespacedName{
```
Comment (nirs) [inline controllers/drplacementcontrol_controller_test.go:695]: Since this is used only in the context of this small helper, and we look up and update one resource, we can use `key` instead of `drpclookupkey`.

This also works for other helpers you modified in this commit.

File: controllers/drplacementcontrol_controller_test.go [hunk 6]
```diff
@@ -1663,27 +1663,15 @@ func deleteDRPolicySync() {
...
-
-		break
+	latestDRCluster := getLatestDRCluster(cluster)
+	if manual {
+		latestDRCluster.Spec.ClusterFence = rmn.ClusterFenceStateManuallyFenced
+	} else {
+		latestDRCluster.Spec.ClusterFence = rmn.ClusterFenceStateFenced
 	}
 
+	updateDRClusterParameters(latestDRCluster)
```
Comment (nirs) [inline controllers/drplacementcontrol_controller_test.go]: Much nicer like this.

File: controllers/drplacementcontrol_controller_test.go (hunk 1 above)
Comment (nirs) [reply controllers/drplacementcontrol_controller_test.go:1686]: Looks good now when we have the same change.

**Review by nirs — COMMENTED** [review]
I'm not sure about the last commit - drclusterConditionExpectEventually is very confusing and hard to follow, but it is better than repeating the complex loops.

We can improve the helper later.

File: controllers/drplacementcontrol_controller_test.go [hunk 7]
```diff
@@ -691,32 +690,28 @@ func deleteNamespaceMWsFromAllClusters(namespace string) {
...
-	}
+		return k8sClient.Update(context.TODO(), latestDRPC)
+	})
 
-	Expect(localRetries).ToNot(Equal(updateRetries))
+	Expect(retryErr).NotTo(HaveOccurred())
 
 	Eventually(func() bool {
-		latestDRPC := getLatestDRPC()
+		latestDRPC = getLatestDRPC()
```
[the remark this answers is not in the document]
Comment (ELENAGER) [reply controllers/drplacementcontrol_controller_test.go:716]: latestDRPC was declared before, no need to do it once again, only to rewrite.

**Review by ShyamsundarR — APPROVED** [review]

File: controllers/drplacementcontrol_controller_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ShyamsundarR) [reply controllers/drplacementcontrol_controller_test.go:716]: Considering this a nit hence merged

File: controllers/drplacementcontrol_controller_test.go (hunk 5 above)
Comment (ShyamsundarR) [reply controllers/drplacementcontrol_controller_test.go:695]: Considering this a nit hence merged
